ideascube/serveradmin/catalog.py

```diff
@@ -0,0 +1,507 @@
+                data = f.read(8388608)
+
+                if not data:
+                    break
+
+                sha.update(data)
+
+        return sha.hexdigest() == sha256sum
+
+    def _fetch_package(self, package):
+        def _progress(*args):
+            self._progress(' {}'.format(package.id), *args)
+
+        filename = '{0.id}-{0.version}'.format(package)
+
+        for cache in self._package_caches:
+            path = os.path.join(cache, filename)
+
+            if os.path.isfile(path):
+                if self._verify_sha256(path, package.sha256sum):
+                    return path
+
+                try:
+                    # This might be an incomplete download, try finishing it
+                    urlretrieve(
+                        package.url, path, sha256sum=package.sha256sum,
+                        reporthook=_progress)
+                    return path
+
+                except DownloadError as e:
+                    # File was too busted, could not finish the download
+                    if e.args[0] is DownloadCheck.checksum_mismatch:
+                        msg = 'Downloaded file has invalid checksum\n'
+
+                    else:
+                        msg = 'Error downloading the file: {}\n'.format(e)
+
+                    sys.stderr.write(msg)
+                    sys.stderr.flush()
+                    os.unlink(path)
+
+        path = os.path.join(self._local_package_cache, filename)
+        urlretrieve(
+            package.url, path, sha256sum=package.sha256sum,
+            reporthook=_progress)
+
+        return path
+
+    def list_installed(self, ids):
+        pkgs = self._get_packages(
+            ids, self._catalog['installed'], fail_if_no_match=False)
+        return sorted(pkgs, key=attrgetter('id'))
+
+    def list_available(self, ids):
+        pkgs = self._get_packages(
+            ids, self._catalog['available'], fail_if_no_match=False)
+        return sorted(pkgs, key=attrgetter('id'))
+
+    def list_upgradable(self, ids):
+        pkgs = []
+
+        for ipkg in self._get_packages(
+                ids, self._catalog['installed'], fail_if_no_match=False):
+            upkg = self._get_package(ipkg.id, self._catalog['available'])
+
+            if ipkg != upkg:
+                pkgs.append(upkg)
+
+        return sorted(pkgs, key=attrgetter('id'))
+
+    def install_packages(self, ids):
+        used_handlers = {}
+
+        for pkg in self._get_packages(ids, self._catalog['available']):
+            if pkg.id in self._catalog['installed']:
+                sys.stderr.write('{0.id} is already installed\n'.format(pkg))
+                sys.stderr.flush()
+                continue
+
+            download_path = self._fetch_package(pkg)
+            handler = self._get_handler(pkg)
+            sys.stdout.write('Installing {0.id}\n'.format(pkg))
+            handler.install(pkg, download_path)
+            used_handlers[handler.__class__.__name__] = handler
```

Comment:
Not that a big problem, but we are creating a new handler at each loop even if we were using the same on previous loop. Given that the `commit` is not instance specific, that's fine, but that *may* be a trap in the future if some handler would have some specific bits in their `commit` step.